evm/p2p/les.py

```diff
@@ -6,14 +18,159 @@
+
+    def handle(self, proto, data):
+        decoded = self.decode(data)
+        # FIXME: Should ask only for block headers we don't have yet.
+        proto.send_get_block_headers(decoded['headNumber'])
+        return decoded
+
+
+class GetBlockHeadersQuery(rlp.Serializable):
+    fields = [
+        # TODO: It should be possible to specify the block either by its number or hash, but
+        # for now only the number is supported.
+        ('block', sedes.big_endian_int),
+        ('maxHeaders', sedes.big_endian_int),
+        ('skip', sedes.big_endian_int),
+        ('reverse', sedes.big_endian_int),
+    ]
+
+
+class GetBlockHeaders(Command):
+    _cmd_id = 2
+    structure = [
+        ('request_id', sedes.big_endian_int),
+        ('query', GetBlockHeadersQuery),
+    ]
+
+
+class BlockHeaders(Command):
+    _cmd_id = 3
+    structure = [
+        ('request_id', sedes.big_endian_int),
+        ('buffer_value', sedes.big_endian_int),
+        ('headers', sedes.CountableList(BlockHeader)),
+    ]
 
     def handle(self, proto, data):
-        proto.logger.debug("Got LES/Status msg")
+        decoded = self.decode(data)
+        last_header_seen = decoded['headers'][0].block_number
+        proto.logger.debug("Last block header received: {}".format(last_header_seen))
+        return decoded
 
 
 class LESProtocol(Protocol):
     name = b'les'
     version = 1
-    _commands = [Status]
-    # FIXME: Need to find out the correct value for this
-    cmd_length = 21
+    _commands = [Status, Announce, BlockHeaders]
+    _req_id = 0
+    max_headers = 10
+    handshake_msg_type = Status
+    cmd_length = 15
+
+    def send_handshake(self):
+        resp = {
+            'protocolVersion': self.version,
+            # FIXME: Need a Chain instance to get the values below from.
+            'networkId': 3,
+            'headTd': GENESIS_DIFFICULTY,
+            'headHash': ZERO_HASH32,
+            'headNum': GENESIS_BLOCK_NUMBER,
+            'genesisHash': decode_hex(
+                '0x41941023680923e0fe4d74a34bdac8141f2540e3ae90623718e47d66d1ca4a2d'),
+        }
+        cmd = Status(self.cmd_id_offset)
+        self.send(*cmd.encode(resp))
+        self.logger.debug("Sending LES/Status msg: {}".format(resp))
+
+    def process_handshake(self, decoded_msg):
+        # TODO: Possibly disconnect if any of
+        # 1. remote doesn't serve headers
+        # 2. genesis hash does not match
+        pass
+
+    def next_req_id(self):
+        self._req_id += 1
+        return self._req_id
+
+    def send_get_block_headers(self, block_number, reverse=True, max_headers=max_headers):
+        """Send a GetBlockHeaders msg to the remote.
+
+        This requests that the remote send us up to max_headers, starting from block_number if
+        reverse is False or ending at block_number if reverse is True.
+        """
+        cmd = GetBlockHeaders(self.cmd_id_offset)
+        req_id = self.next_req_id()
+        skip = 0
```

Comment:
Maybe a comment explaining what `skip` does here?  This is confusing to me since I don't know the protocol very well at this point.

Reply:
Done. It's what python APIs call step. I could rename it but I think it's better to use the same name used in the spec, with a comment